### ISSUE TITLE
support for istio distributed tracing

### ISSUE THREAD

**Issue by salanfe** [go-kit/kit#977]
Hi guys,

I've been using go-kit for some times now (high five for the amazing work !) and have been developing micro-services on Kubernetes and istio where istio tracing is leveraged. If interested in supporting istio tracing through simple header propagation ([istio doc](https://istio.io/faq/distributed-tracing/#how-to-support-tracing)), I have working code that I would be happy to push through a pull request.

Basically the code implements the function `func ServerBefore(before ...RequestFunc) ServerOption` for HTTP and gRPC

defined here
* https://github.com/go-kit/kit/blob/master/transport/http/server.go#L51
* https://github.com/go-kit/kit/blob/master/transport/grpc/server.go#L60

so that header values can be saved in the `context`.

Is it something you see being pushed to your code base ? Let me know.
Cheers

**Comment by basvanbeek** [comment]
Specifically tailored to Istio I think it does not provide a whole lot of value. I do see where one could have a generic "baggage" like header propagation middleware for clients and servers with the ability to specifically whitelist or blacklist which headers to allow or deny for propagation.

By the way if you want to have Go kit interop in distributed tracing you can already do so. Go kit supports Zipkin, OpenTracing and OpenCensus instrumentation thus you can use these to already propagate traces (and even augment traces with local in process spans).

**Comment by salanfe** [comment]
Hi, thanks for your reply. I understand, makes sense. Today, the trend is to offload middlewares to the service mesh as much as possible : middlewares like

* JWT verification
* request tracing
* circuit breakers
* fault injections
* etc (see istio features)

however tracing or accessing custom claims in JWT (for example) still require to propagate requests' headers to downstream services. This is two areas where header propagation has been implemented in our backends. So your suggestion of a generic "baggage" middleware for header propagation makes sense to me.

**Comment by lewsmith** [comment]
@salanfe, it's been a while, but do you have an example repo showing your usage of `ServerBefore/ServerOption` for istio header propagation?

**Comment by intel352** [comment]
@peterbourgon you noted this as help wanted, what's the expected direction here?
